### ISSUE TITLE
想请问一下怎么复现第一篇文章的下游任务结果呢？

### ISSUE THREAD

**Issue by lain117** [chuanyang-Zheng/CAPE#2]
使用代码里的eval.py有不少报错

**Comment by chuanyang-Zheng** [comment]
我们用的是这里的代码  https://github.com/google-deepmind/randomized_positional_encodings

如果有其他问题，随时交流

**Comment by lain117** [comment]
> 我们用的是这里的代码 https://github.com/google-deepmind/randomized_positional_encodings
> 
> 如果有其他问题，随时交流

感谢您的回复，祝科研顺利

**Comment by lain117** [comment]
> 我们用的是这里的代码 https://github.com/google-deepmind/randomized_positional_encodings
> 
> 如果有其他问题，随时交流

还想请教的一点是，使用您的代码训练保存的neox框架的checkpoints，是如何转移到这个项目的jax框架下进行eval的呢

**Comment by chuanyang-Zheng** [comment]
在下游任务，是直接在这里的代码训练的https://github.com/google-deepmind/randomized_positional_encodings
这里已经有了alibi kerple rope等等，只要用jax的linear mapping处理alibi kerple就可以实现dape，如果有代码问题，可以查看询问deepseek 让它帮你从pytorch变成jax一下就行了

**Comment by lain117** [comment]
> 在下游任务，是直接在这里的代码训练的https://github.com/google-deepmind/randomized_positional_encodings 这里已经有了alibi kerple rope等等，只要用jax的linear mapping处理alibi kerple就可以实现dape，如果有代码问题，可以查看询问deepseek 让它帮你从pytorch变成jax一下就行了

明白了，感谢您的回复

**Comment by lain117** [comment]
> 在下游任务，是直接在这里的代码训练的https://github.com/google-deepmind/randomized_positional_encodings 这里已经有了alibi kerple rope等等，只要用jax的linear mapping处理alibi kerple就可以实现dape，如果有代码问题，可以查看询问deepseek 让它帮你从pytorch变成jax一下就行了
还有一些困惑要打扰您，获取了arxiv和books3数据集之后，我运行您的代码训练了cape_kerple，得到的ppl是4x10^1，对比论文中的数据缺少了一个数量级，在book3数据集上得到的结果也是2x10^2多，数字吻合但少了一个数量级，请问这是正确的吗。以及我在下游任务的复现中，比如评估rotray的R任务even_pairs，modular，parity_check，cycle_navigation；准确率分别是0.5，0.2，0.5，0.2，除了第一个只是一半其他的基本能对上，我用别的方法也是准确率接近1的任务都只有五十上下，准确率比较低的任务都是吻合的。请问这两个实验结果问题是有哪里的setting有改动我没注意到吗，麻烦您了

**Comment by chuanyang-Zheng** [comment]
对于gpt neox的结果，你运行的是哪个position encoding？对于randomized pe的代码，你要看下有没有开randomized position encoding这个tricks，如果没有开，那rope效果就是很差的 可以参考下这里https://arxiv.org/abs/2305.16843 Table 2

**Comment by chuanyang-Zheng** [comment]
我看了下，你可能需要将config里面的pipeline parallel改成0，gpt neox的代码更新了。以及，你是不是用了gradient accumulation，你检查下training loss和对应长度的evaluation loss是不是几乎一样的

**Comment by lain117** [comment]
> 我看了下，你可能需要将config里面的pipeline parallel改成0，gpt neox的代码更新了。以及，你是不是用了gradient accumulation，你检查下training loss和对应长度的evaluation loss是不是几乎一样的

对于randomized我了解了，还想请问的是您的方法开启了这个tricks吗。然后是关于neox的结果，我运行的是是125M_kerple_c这个方法。确实是training loss和eval loss基本一致，不过我没有在config里看到梯度累计（gradient accumulation）的参数，同时我也是单卡进行训练的（我使用的服务器linux内核较低并行训练会导致线程挂起），没有修改pipeline parallel仍然会影响我的结果吗

**Comment by chuanyang-Zheng** [comment]
是的，方法是用了randomized的，你自己的setting可以选择开或者不开，只要实验对比公平就行。

对于外推结果的问题，我想知道对于book数据集，training loss是多少，validation loss以及对应的长度是多少？

**Comment by lain117** [comment]
下游任务的问题感谢解答。对于外推结果的问题，我这边还只train了arxiv数据集，我在log里拿一个kerple_cape的五万步的例子（train的长度是512）结果是这样的：steps: 50000 loss: 3.9247 iter time (s): 8.597 samples/sec: 7.444； validation results at Seq_Length 512: the end of training for val data | lm_loss value: 3.814293E+00 | lm_loss_ppl value: 4.534468E+01；validation results at Seq_Length 1024: the end of training for val data | lm_loss value: 3.890501E+00 | lm_loss_ppl value: 4.893539E+01；lidation results at Seq_Length 2048: the end of training for val data | lm_loss value: 3.820705E+00 | lm_loss_ppl value: 4.563639E+01 |；lidation results at Seq_Length 4096: the end of training for val data | lm_loss value: 3.879429E+00 | lm_loss_ppl value: 4.839655E+01 ；由于我这一个服务器的显存不够，暂时去掉了8192的推理步骤

**Comment by chuanyang-Zheng** [comment]
外推相比于训练没有明显上升，说明外推没问题。应该是训练不充分导致的，我之前用的是八卡训练，现在单卡训练只有之前训练的八分之一

你接下来这样做
1.https://github.com/chuanyang-Zheng/DAPE/blob/main/megatron/training.py#L1822
改成forward_step_func=forward_step

2.在config里面加上 ”gradient_accumulation_steps“: 4
3.pipe-parallel-size改成0
4.eval-interval改成100进行快速debug

**Comment by chuanyang-Zheng** [comment]
结论: 外推 loss相比训练loss有下降，说明可以进行外推。效果比文章中包含要差，因为现在单卡训练，训练数据只有文章中八分一。
你可以用单卡进行快速迭代研究，只要保持实验是公平的就行

**Comment by lain117** [comment]
> 结论: 外推 loss相比训练loss有下降，说明可以进行外推。效果比文章中包含要差，因为现在单卡训练，训练数据只有文章中八分一。 你可以用单卡进行快速迭代研究，只要保持实验是公平的就行

了解了。非常感谢您的指导！
